ibis/expr/rules.py

```diff
@@ -275,22 +275,22 @@ def output_shape(self):
 # TODO(kszucs): might just use bounds instead of actual literal values
 # that could simplify interval binop output_type methods
 # TODO(kszucs): pre-generate mapping?
-def _promote_numeric_binop(exprs, op):
-    bounds, dtypes = [], []
-    for arg in exprs:
-        dtypes.append(arg.type())
-        if hasattr(arg.op(), 'value'):
-            # arg.op() is a literal
-            bounds.append([arg.op().value])
-        else:
-            bounds.append(arg.type().bounds)
+def _promote_integral_binop(exprs, op):
```

Comment:
To support unsigned addition, I had to modify the promotion algorithm here a bit. I don't believe there should be any change in behavior for signed integer/literal addition, but it now works for (some) cases of unsigned integers.

Note that this promotion rule assumes that the backing compute engine promotes on overflow rather than wraps or saturates. This means that `uint32 + 1` always promotes to `uint64`, even if all values in the column would still fit in a `uint32`. It also means that `uint64 - 1` currently errors, since there's no ibis type that can fit `[-1, uint64.max - 1]`. `uint32 - 1` also currently errors, but could be made to work (I guess it should be an `int64`). I'm not really sure what the expected behavior should be here, suggestions welcome.

Reply:
I think the result type of `+/- 1` depends on what the inferred type of `1` is. If we infer `1` as any kind of signed type then I think we have to promote to the largest signed integer type `int64`. The only case where that might be different is in the case of two literal operands and I think that would complicate the promotion logic quite a bit since we'd have to look not only at the value but also at the operation type.

Looking at julia and numpy, both promote to `int64` in the case of `uint32(0) +/- 1`:

numpy:

```
In [3]: import numpy as np

In [4]: type(np.uint32(0) - 1)
Out[4]: numpy.int64

In [5]: type(np.uint32(0) + 1)
Out[5]: numpy.int64
```

Julia:
```
julia> typeof(UInt32(0) - 1)
Int64

julia> typeof(UInt32(0) + 1)
Int64
```

In the case where `1` is an unsigned integer then both numpy and Julia wrap.

In the database world things are a bit dicey:

DuckDB:

```
D select typeof(cast(0 as ubigint) - 1) as x;
┌────────┐
│   x    │
├────────┤
│ DOUBLE │
└────────┘
```

It looks like MySQL supports `UNSIGNED` in DDL, but I'm not sure if there's a syntax for using these types at runtime, for example, in a `CAST` expression. The naive approach doesn't seem to work:

```
MariaDB [ibis_testing]> select cast(1 as int unsigned);
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'unsigned)' at line 1
```

All this is to say I think the behavior should be:

* `unsigned ± signed -> int64`
* `unsignedX ± unsignedY -> max(unsignedX, unsignedY)` (this may or may not error depending on the system, DuckDB errors if the result is out of range for type, pandas/numpy wrap)

Reply:
> Looking at julia and numpy, both promote to int64 in the case of uint32(0) +/- 1:

NumPy's promotion rules differ depending on if the item is an array or a scalar. In the case of unsigned-array + python-int, the dtype stays the same and addition wraps.

```python
In [8]: (np.uint32(1) + 1).dtype
Out[8]: dtype('int64')

In [9]: (np.array([1], dtype=np.uint32) + 1).dtype
Out[9]: dtype('uint32')
```

Since duckdb tries to do exact math when possible (promoting to a larger type on overflow), trying to emulate that intuitively in ibis would be nice.

So I guess I agree with your behavior described, except that I think in the case of adding with a positive python integer, the signedness shouldn't change, so `uint64 + 1` is still a `uint64`, but `uint64 - 1` would be an `int64`, as would `uint64 + ibis.literal(1, "int64")` (since that's an explicitly typed integer).

Reply:
That sounds reasonable to me.